smarts/core/traffic_history.py

```diff
@@ -0,0 +1,124 @@
+# Copyright (C) 2020. Huawei Technologies Co., Ltd. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+
+from cached_property import cached_property
+from contextlib import nullcontext, closing
+from functools import lru_cache
+import random
+import sqlite3
+
+
+class TrafficHistory:
+    def __init__(self, db):
+        self._db = db
+        self._db_cnxn = None
+
+    def connect_if_not(self):
+        if not self._db_cnxn:
+            self._db_cnxn = sqlite3.connect(self._db)
+
+    def close_db(self):
+        if self._db_cnxn:
+            self._db_cnxn.close()
+            self._db_cnxn = None
+
+    def _query_val(self, query, params=(), result_type=float):
+        with nullcontext(self._db_cnxn) if self._db_cnxn else closing(
+            sqlite3.connect(self._db)
+        ) as dbcnxn:
+            cur = dbcnxn.cursor()
+            cur.execute(query, params)
+            row = cur.fetchone()
+            cur.close()
+        if not row:
+            return None
+        return row if result_type is tuple else result_type(row[0])
+
+    def _query_list(self, query, params=()):
+        with nullcontext(self._db_cnxn) if self._db_cnxn else closing(
+            sqlite3.connect(self._db)
+        ) as dbcnxn:
+            cur = dbcnxn.cursor()
+            for row in cur.execute(query, params):
+                yield row
+            cur.close()
+
+    @cached_property
+    def lane_width(self):
+        return self._query_val("SELECT value FROM Spec where key='map_net.lane_width'")
+
+    @cached_property
+    def target_speed(self):
+        return self._query_val("SELECT value FROM Spec where key='speed_limit_mps'")
+
+    @lru_cache(maxsize=32)
+    def vehicle_last_seen_time(self, vehicle_id):
```

Comment:
Perhaps `vehicle_final_exit_time`? I was wondering if this was referring to the most recent time the vehicle was seen.